dan-riley/marble_mapping
# Feature/stair data (#1)

* initial implementation, seems to be working

* tweaking launchfiles to enable subt team merging

* enable multiple namespacing in launchfiles, removed unnec comment

* tweaked launchfiles for team deployement and merging, improved traversability merging performance

* added build flag for enabling traversability info, seems to be working

* added traversability and sharing enable/disable param

* source RoughOctomap features from rough_octomap package rather than modified version of octomap

* added base_frame_id arg so that octomap radius follows vehicle

* removed unnec lines from launchfile

* fixed bug with wrong include

* cleaned up launchfiles, added height color to markers when traversability disabled

* fixed hole in map bug, removed nan filter for traversability intensity fields (too allow input of PointXYZ clouds even with traversability enabled, considered handling nan inputs with ifs but will handle in RoughOctomap class instead), added clear merged_tree (along with octree) on reset, tweaked launchfiles (added respawns)

* see prev

* working on integrating into sim, i think last merge broke something

* fixed merge conflict which added unnec mutex lock and freezing

* merged rough_octomap into scan_plan for sim

* fixed bug where merged_octree roughness was double averaged compared to octree, modded pipeline to use mixed clouds instead of concat'ing

* implemented voxel grid filter, passthrough filter, and outlier removal prior to terrain classification for noisy picoflexx points, and disable downsampling/region-growing in curvy_terrain_mapper

* switched to integrateNodeRough rather than averageNodeRough, plus fixed double-averaging bug

* fixed merge conflicts

* Successful Edgar deployment 2021-06-17 (H01 exploration and terrain): 1 hour exploration test scan_plan with traversability mapping, only got stuck 3x on RR tracks, needed to backup 25-50cm before letting it continue following path. Targeted terrain testing by the sloped/staired hill in the back of Edgar, avoided it successfully. Targeted terrain testing by muddy tired tracked hills, it got stuck up there and got programmatically stuck before getting to the cliff, need to analyze bag data. Overall, very happy with the performance, traversability is tuned well, except for perhaps it was slightly too aggressive in the muddy tire track area.

* impl stair logodds mapping, need to test

* fixed bug with seg fault on insert stair scan, added stair text markers

* fixed issue with no downpressure on stair data

* started impl of stair edge extraction

* finished edge extraction/pub

* Mike M put on H02 to test open loop detection

* added angle check to stairs, fixed bug in which stairs only updated on merged map

* revamped how stair clouds are consumed to prevent seg fault, may not be best way tho

* tried out new filtering techniques, added viz marker deletion, reduced clustering tolerance

* modified how mapping consumes stair data, fixed access segfaults

* minor bug

* had to enable marker_array to get stair viz

* added near stair indicator, moved stair stuff into separate timer

* added stairs probability params

* minor

* Optimized insertStairScan and other cleanup

* Performance update

* added base frame to launch args

* set near stair indicator radius threshold to be based only on x,y (rather than x,y,z)

* set near stair indicator radius threshold to be based on x,y (rather than x,y,z)

* Added fixed agent colors via launch file

* Another base station crash fix

Co-authored-by: mohradza <mioh0887@colorado.edu>
Co-authored-by: H01 <marblerobots@gmail.com>
Co-authored-by: marble-spot <D01@marble.com>
Co-authored-by: Dan Riley <dan.riley@colorado.edu>
